kr_mesh_visualization/src/mesh_visualization.cpp

```diff
@@ -1,83 +1,100 @@
-#include <geometry_msgs/PoseStamped.h>
-#include <geometry_msgs/PoseWithCovarianceStamped.h>
-#include <nav_msgs/Odometry.h>
-#include <ros/ros.h>
-#include <topic_tools/shape_shifter.h>
-#include <visualization_msgs/Marker.h>
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
+#include "nav_msgs/msg/odometry.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "visualization_msgs/msg/marker.hpp"
 
-static std::string mesh_resource, new_frame_id;
-static ros::Publisher pub_vis;
-static double color_r, color_g, color_b, color_a;
-static double scale_x, scale_y, scale_z;
-
-static void publishMarker(const std::string &frame_id, const geometry_msgs::Pose &pose)
+class MeshVisualizationNode : public rclcpp::Node
 {
-  visualization_msgs::Marker marker;
-  marker.header.frame_id = (new_frame_id == "") ? frame_id : new_frame_id;
-  marker.header.stamp = ros::Time();  // time 0 so that the marker will be
-                                      // displayed regardless of the current time
-  marker.ns = ros::this_node::getName();
-  marker.id = 0;
-  marker.type = visualization_msgs::Marker::MESH_RESOURCE;
-  marker.action = visualization_msgs::Marker::ADD;
-  marker.pose = pose;
-  marker.scale.x = scale_x;
-  marker.scale.y = scale_y;
-  marker.scale.z = scale_z;
-  marker.color.a = color_a;
-  marker.color.r = color_r;
-  marker.color.g = color_g;
-  marker.color.b = color_b;
-  marker.mesh_resource = mesh_resource;
-  pub_vis.publish(marker);
-}
+ public:
+  MeshVisualizationNode() : Node("kr_mesh_visualization")
+  {
+    this->declare_parameter<std::string>("mesh_resource",
+                                         std::string("package://kr_mesh_visualization/mesh/hummingbird.mesh"));
+    this->declare_parameter<double>("color/r", 1.0);
```

Comment:
This seems a bit inefficient. You generally declare colors by rgb tuples `[255, 0, 0]`. Same for the scale.

Reply:
True, but I simply followed the structure of the original code. Tried to keep it similar.